lightning/src/routing/router.rs

```diff
@@ -333,84 +592,262 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, targ
 		}
 
-		match network.get_nodes().get(&pubkey) {
-			None => {},
-			Some(node) => {
-				add_entries_to_cheapest_to_target_node!(node, &pubkey, lowest_fee_to_node);
-			},
+		// Step (3).
+		// Stop either when recommended value is reached,
+		// or if during last iteration no new path was found.
+		// In the latter case, making another path finding attempt could not help,
+		// because we deterministically terminate the search due to low liquidity.
+		if already_collected_value_msat >= recommended_value_msat || !found_new_path {
+			break 'paths_collection;
+		}
+	}
+
+	// Step (4).
+	if payment_paths.len() == 0 {
+		return Err(LightningError{err: "Failed to find a path to the given destination".to_owned(), action: ErrorAction::IgnoreError});
+	}
+
+	if already_collected_value_msat < final_value_msat {
+		return Err(LightningError{err: "Failed to find a sufficient route to the given destination".to_owned(), action: ErrorAction::IgnoreError});
+	}
+
+	// Sort by total fees and take the best paths.
+	payment_paths.sort_by_key(|path| path.get_total_fee_paid_msat());
+	if payment_paths.len() > 50 {
+		payment_paths.truncate(50);
+	}
+
+	// Draw multiple sufficient routes by randomly combining the selected paths.
+	let mut drawn_routes = Vec::new();
+	for i in 0..payment_paths.len() {
+		let mut cur_route = Vec::<PaymentPath>::new();
+		let mut aggregate_route_value_msat = 0;
+
+		// Step (5).
+		// TODO: real random shuffle
+		// Currently just starts with i_th and goes up to i-1_th in a looped way.
+		let cur_payment_paths = [&payment_paths[i..], &payment_paths[..i]].concat();
+
+		// Step (6).
+		for payment_path in cur_payment_paths {
+			cur_route.push(payment_path.clone());
+			aggregate_route_value_msat += payment_path.get_value_msat();
+			if aggregate_route_value_msat >= final_value_msat {
+				// Last path likely overpaid. Substract it from the most expensive
+				// (in terms of proportional fee) path in this route and recompute fees.
+				// This might be not the most economically efficient way, but fewer paths
+				// also makes routing more reliable.
+				let mut overpaid_value_msat = aggregate_route_value_msat - final_value_msat;
+
+				// First, drop some expensive low-value paths entirely if possible.
+				// Sort by value so that we drop many really-low values first.
+				cur_route.sort_by_key(|path| path.get_value_msat());
+				// We should make sure that at least 1 path left.
+				let mut paths_left = cur_route.len();
+				cur_route.retain(|path| {
+					if paths_left == 1 {
+						return true
+					}
+					let mut keep = true;
+					let path_value_msat = path.get_value_msat();
+					if path_value_msat <= overpaid_value_msat {
+						keep = false;
+						overpaid_value_msat -= path_value_msat;
+						paths_left -= 1;
+					}
+					keep
+				});
+
+				if overpaid_value_msat == 0 {
+					break;
+				}
+
+				assert!(cur_route.len() > 0);
+
+				// Step (7).
+				// Now, substract from the most-expensive path the remaining value.
```

Comment:
Should say the overpaid value ? I understand remaining value by what we're going to flow.